PoS: Custom buy button text per product

dennisreimann opened the pull request:
Closes #2232.

I saw the bounty by @MaxHillebrand in the chat and did this – only realizing afterwards that @bolatovumar already discussed this with Max and wanted to take it on. As I didn't intend to get in the way I'd suggest Umar reviews this and we split the bounty if that's okay too.

## Editor

Works with regular text too, this example uses the `{0}` to be replaced by the amount, which is supported on a store level too. 

![editor](https://user-images.githubusercontent.com/886/108340489-ec85d300-71d8-11eb-8e3e-66220dd35bf3.png)

## Static View

![static](https://user-images.githubusercontent.com/886/108340624-14753680-71d9-11eb-8757-27c05ad18bfd.png)

## Cart View

![cart](https://user-images.githubusercontent.com/886/108340659-1b03ae00-71d9-11eb-84e3-951bcbbe6891.png)

Kukks reviewed (changes requested):
Need to take into account crowdfund app too as it uses the same templating

dennisreimann commented:
@Kukks Thanks for the pointer, added.

![crowdfund](https://user-images.githubusercontent.com/886/108351179-7471da00-71e5-11eb-9369-d090c0821625.png)

MaxHillebrand commented:
You're a treasure @dennisreimann, thanks for picking it up so fast :) 

Yes, I'm happy to include reviewers of the PR in the bounty.

Kukks commented on BTCPayServer/Models/AppViewModels/ViewPointOfSaleViewModel.cs:
```diff
@@ -17,6 +17,7 @@ public class ItemPrice
             public ItemPrice Price { get; set; }
             public string Title { get; set; }
             public bool Custom { get; set; }
+            public string BuyButtonText { get; set; }
```
`BuyButtonText` or simply `ButtonText`?

dennisreimann replied:
I thought about this too and was torn: The upper level prop is called `CustomButtonText` – one might argue we should match this, but I wanted to avoid confusion and used the buy-prefix to be specific … so yeah, let's discuss this.

bolatovumar replied:
I think `BuyButtonText` is more accurate even though it's more verbose.

Kukks reviewed (approved):
LGTM, just a small quyestion if the prop name should be simpler or not but nothing major

pavlenex reviewed (changes requested):
If `custom price` is set to `yes` the text you add in the button name field doesn't take effect. If it's desired behavior, we should probably hide the `Buy Button Text` field, if it's not desired, the name in the `Buy Button Text` should take effect even if custom price is set to `yes`.

bolatovumar commented:
> If `custom price` is set to `yes` the text you add in the button name field doesn't take effect. If it's desired behavior, we should probably hide the `Buy Button Text` field, if it's not desired, the name in the `Buy Button Text` should take effect even if custom price is set to `yes`.
> 
> ![Screenshot from 2021-02-19 17-23-46](https://user-images.githubusercontent.com/36959754/108531850-75833400-72d7-11eb-8bec-6e47c833c4fe.png)
> 
> ![Screenshot from 2021-02-19 17-26-22](https://user-images.githubusercontent.com/36959754/108531974-99df1080-72d7-11eb-8122-d276d7cc11d1.png)

Yeah, I think if you specify custom button text it should apply no matter what you select for "Custom price". The user who puts in custom button text probably doesn't care what is selected in the "Custom price" dropdown when they set the button text, they just want this text to appear no matter what.

dennisreimann commented:
Fixed, also updated the display so that it works with long and short text …

![custom-price](https://user-images.githubusercontent.com/886/108679316-9ffa0a80-74ec-11eb-84d6-fc52011cecd3.png)

pavlenex reviewed (approved):
tACK.

bolatovumar reviewed (commented):
I tested this branch locally on my machine and noticed a couple of issues:

1. If I add a custom text to the first item in the default POS app (green tea) and then update it again right after all OTHER items get the old text for the first item after the second update (see screenshot). I can reproduce this consistently.
   - ![Screen Shot 2021-02-22 at 6 59 50 PM](https://user-images.githubusercontent.com/1934678/108797727-baf66a00-7540-11eb-9ff5-c4a1ea744f3f.png)
 2. If I add text that's too long the button overspills. This one is probably not a big deal as I don't expect people will add long text in these buttons
    - ![Screen Shot 2021-02-22 at 7 01 36 PM](https://user-images.githubusercontent.com/1934678/108797770-deb9b000-7540-11eb-805c-adbf668da645.png)

dennisreimann commented:
@bolatovumar good spot, it's fixed now. As far as I read it, we had the same problem before with other properties (price, title, description and custom) because they weren't reset before parsing a new item. Fixed in 32370545cbcc2d7a3141a6ab3fd19ecb94ccc8d6.

bolatovumar commented on BTCPayServer.Tests/SeleniumTests.cs:
```diff
@@ -478,12 +478,21 @@ public async Task CanCreateAppPoS()
                 s.Driver.FindElement(By.Id("SelectedStore")).SendKeys(storeName);
                 s.Driver.FindElement(By.Id("Create")).Click();
                 s.Driver.FindElement(By.Id("DefaultView")).SendKeys("Cart");
+                s.Driver.FindElement(By.CssSelector(".template-item:nth-of-type(1) .btn-primary")).Click();
```
Wouldn't it be better to use some `test-id` attribute instead so if a new button is added or something you don't have to update this test?

dennisreimann replied:
In general I agree. This one selects the first article in the editor – as it is a dynamic list I thought it'd be ok this way too.

bolatovumar commented on BTCPayServer/Views/Apps/TemplateEditor.cshtml:
```diff
@@ -113,11 +113,17 @@
+                            <div class="form-row">
+                                <div class="col">
+                                    <label>Buy Button Text</label>
+                                    <input type="text" id="BuyButtonText" class="form-control mb-2" v-model="editingItem.buyButtonText" ref="txtBuyButtonText"/>
```
`ref` attribute seems weird. `txtBuyButtonText` <-- looks like `txt` and `Text` are meant to represent text and are repeated?

dennisreimann replied:
It's the same as for the other input fields – I think the `txt` part refers to the input being a text input. Can change it, but think we should then change the other field refs too to be consistent.

bolatovumar commented on BTCPayServer/Views/AppsPublic/PointOfSale/Cart.cshtml:
```diff
@@ -253,7 +253,7 @@
                         <div class="card-footer pt-0 bg-transparent border-0">
 
-                            <span class="text-muted small">@Model.ButtonText.Replace("{0}",item.Price.Formatted).Replace("{Price}",item.Price.Formatted)</span>
+                            <span class="text-muted small">@((item.BuyButtonText ?? Model.ButtonText).Replace("{0}",item.Price.Formatted).Replace("{Price}",item.Price.Formatted))</span>
```
Can we confirm with @MaxHillebrand that we also want to modify the button text in the cart?

MaxHillebrand replied:
Yes please. It makes sense to have this custom too, for the same reasons as the other buttons.

bolatovumar reviewed (approved):
Tested the updated branch and couldn't find any issues.

dennisreimann commented:
Fix for the Selenium test contained in #2310. Can rebase this one if the other PR gets merged first.

MaxHillebrand commented on BTCPayServer.Tests/README.md:
```diff
@@ -26,6 +26,21 @@ You can also generate blocks:
+LND Connection string example: 
+type=lnd-rest;server=https://127.0.0.1:8084/;macaroonfilepath="local path to admin.macaroon on your computer, without these quotes";allowinsecure=true
+
+Now you can create a lightning invoice on BTCPay Server regtest and make a payment through Polar.
```
```suggestion
Now you can create a Lightning invoice on BTCPay Server regtest and make a payment through Polar.
```

minor typo nit for consistency.

dennisreimann replied:
Thanks for fixing! Not sure though why this slipped in there with my rebase … it's unrelated and present on master already.

MaxHillebrand commented:
Thank you @dennisreimann @bolatovumar @Kukks @pavlenex for all you past and present support of free software like BTCPay! Y'all are heroes!!

Some sats have magically found their way into your wallets. :black_heart: :fire:
